llvm/lib/Bitcode/Reader/BitcodeReader.cpp

```diff
@@ -4301,10 +4313,13 @@ Error BitcodeReader::parseModule(uint64_t ResumeBit,
                                  bool ShouldLazyLoadMetadata,
                                  ParserCallbacks Callbacks) {
   // Load directly into RemoveDIs format if LoadBitcodeIntoNewDbgInfoFormat
-  // has been set to true (default action: load into the old debug format).
-  TheModule->IsNewDbgInfoFormat =
-      UseNewDbgInfoFormat &&
-      LoadBitcodeIntoNewDbgInfoFormat == cl::boolOrDefault::BOU_TRUE;
+  // has been set to true and we aren't attempting to preserve the existing
+  // format in the bitcode (default action: load into the old debug format).
+  if (PreserveInputDbgFormat != cl::boolOrDefault::BOU_TRUE) {
```

Comment:
Why isn't this `== cl::boolOrDefault::BOU_TRUE`?

Reply:
The intent is that if we're preserving the input format, then we _don't_ want to update `TheModule->IsNewDbgInfoFormat` according to the rule below. It could all be folded into a boolean expression, but I think that might obfuscate the meaning.

Reply:
Ah I think I see where my confusion comes in, `TheModule->IsNewDbgInfoFormat` is currently always false. Is this patch intended to land after the "set IsNewDbgInfoFormat in Module ctor based on UseNewDbgInfoFormat" patch (alternatively: is your reply above intended to be true after that patch lands, even if this one lands first)?

Otherwise, without the module-ctor-patch, the default behaviour becomes convert all inputs into intrinsic mode (because of the code down on line 6842). Or am I misreading this situation?

Reply:
This doesn't require the ctor change; the behaviour is that:

- If we aren't preserving inputs, then we follow the existing behaviour - we set `IsNewDbgInfoFormat` according to the `UseNewDbgInfoFormat` and `LoadBitcodeIntoNewDbgInfoFormat` flags, and then when we materialize the function we either 1) convert any existing records to intrinsics via the call to `setIsNewDbgInfoFormat` on line 6849, or 2) convert any existing intrinsics to records during autoupgrading, depending on whether we're loading into the new format or not.
- If we are preserving inputs, then we never convert between intrinsics and records at any point - we insert records/intrinsics as we materialize the function, and then we set the `IsNewDbgInfoFormat` flag on the module and its functions once we've determined what format the input was in.